File: packages/sdk-widget/src/audit.ts
import React from "react";
import type { AuditLogEntry } from "./types.js";

/** Write interface injected by the host application. All widget failures are recorded here. */
export interface AuditLog {
  write(entry: AuditLogEntry): void;
}

/** Fallback used when the host has not provided an AuditLogContext.Provider. */
const _noopAuditLog: AuditLog = {
  write: () => undefined,
};

/**
 * React context that carries the host-provided AuditLog implementation.
 * Wrap the canvas root with <AuditLogContext.Provider value={yourLog}>.
 */
export const AuditLogContext = React.createContext<AuditLog>(_noopAuditLog);

/** Convenience hook for reading the current audit log inside a function component. */
export function useAuditLog(): AuditLog {
  return React.useContext(AuditLogContext);
}

File: packages/api-capabilities/src/api_capabilities/_audit.py
from __future__ import annotations

from abc import ABC, abstractmethod

from ._types import AuditLogEntry


class AuditLog(ABC):
    """Write interface injected by the host application. All capabilities API failures
    are recorded here."""

    @abstractmethod
    def write(self, entry: AuditLogEntry) -> None: ...


class NoopAuditLog(AuditLog):
    """Fallback used when the host has not provided an AuditLog implementation."""

    def write(self, entry: AuditLogEntry) -> None:
        pass
